include/ginkgo/core/base/range_accessors.hpp

```diff
@@ -205,6 +631,289 @@ class row_major {
+
+    template <typename FirstType, typename... Indices>
+    static constexpr GKO_ATTRIBUTES ValueType
+    compute(const std::array<ValueType, total_dim> &size,
+            const std::array<ValueType, (total_dim > 1 ? total_dim - 1 : 0)>
+                &stride,
+            FirstType first, Indices &&...)
+    {
+        return GKO_ASSERT(first < size[total_dim - 1]),
+               first * stride[dim_idx - 1];
+    }
+};
+
+/**
+ * Computes the storage index for the given indices with respect to the given
+ * stride array
+ */
+template <typename ValueType, size_type total_dim, typename... Indices>
+constexpr GKO_ATTRIBUTES ValueType compute_index(
+    const std::array<ValueType, total_dim> &size,
+    const std::array<ValueType, (total_dim > 0 ? total_dim - 1 : 0)> &stride,
+    Indices &&... idxs)
+{
+    return index_helper_s<ValueType, total_dim>::compute(
+        size, stride, std::forward<Indices>(idxs)...);
+}
+
+
+template <size_type iter = 1, typename ValueType, size_type N, typename... Args>
+constexpr GKO_ATTRIBUTES
+    std::enable_if_t<(iter == N - 1) && (iter == sizeof...(Args) + 1),
+                     std::array<ValueType, N - 1>>
+    default_stride_array_impl(const std::array<ValueType, N> &size,
+                              Args &&... args)
+{
+    return {{std::forward<Args>(args)..., size[N - 2]}};
+}
+
+template <size_type iter = 1, typename ValueType, size_type N, typename... Args>
+constexpr GKO_ATTRIBUTES std::enable_if_t<(iter < N - 1 || iter == N) &&
+                                              (iter == sizeof...(Args) + 1),
+                                          std::array<ValueType, N - 1>>
+default_stride_array_impl(const std::array<ValueType, N> &size, Args &&... args)
+{
+    return default_stride_array_impl<iter + 1>(
+        size, std::forward<Args>(args)...,
+        detail::mult_dim_upwards_impl<iter>(size));
+}
+
+template <typename ValueType, size_type dimensions>
+constexpr GKO_ATTRIBUTES
+    std::array<ValueType, (dimensions > 0 ? dimensions - 1 : 0)>
+    default_stride_array(const std::array<ValueType, dimensions> &size)
+{
+    return default_stride_array_impl(size);
+}
+
+
+}  // namespace detail_colmajor
+
+
+/**
+ * A bridge between a range and a column-major memory layout.
+ *
+ * Note that this is not completely a 'layout right'. Only the innermost
+ * two dimensions are regarded as defining a column-major matrix, and
+ * the rest of the dimensions are treated identically to \ref row_major.
+ *
+ * You should not try to explicitly create an instance of this accessor.
+ * Instead, supply it as a template parameter to a range, and pass the
+ * constructor parameters for this class to the range (it will forward it to
+ * this class).
+ *
+ * @tparam ValueType  type of values this accessor returns
+ * @tparam Dimensionality  number of dimensions of this accessor
+ */
+template <typename ValueType, size_type Dimensionality>
+class col_major {
```

Comment:
I don't think so.. I need the blocks to be contiguous, not interleaved, at least for now. A true 3D layout_right, for example, would have the (0,0) entry of all blocks first, then the (1,0) entry of all blocks, and so on. That's suitable sometimes, but not what I'm looking for right now. I also think  "row major" and "column major" really only make sense in 2D. But I'm okay with `block_col_major` for this.

Reply:
Row- and column-major do exist for more than 2D (see the wiki page about [Row- and column-major](https://en.wikipedia.org/wiki/Row-_and_column-major_order)). It might be useful to have a full column-major implementation in the future, which is why I would prefer if you would rename it to `block_col_major`.